export/export.py

```diff
@@ -0,0 +1,866 @@
+from executorch.runtime import Runtime, Verification
+from tabulate import tabulate
+from torch import nn
+from torch.ao.quantization import allow_exported_model_train_eval
+from torch.ao.quantization.quantize_pt2e import convert_pt2e, prepare_pt2e
+from torch.export import export_for_training, ExportedProgram
+from torchao.quantization import quantize_
+from torchao.utils import unwrap_tensor_subclass
+
+from .recipe import ExportRecipe
+
+
+class Stage(ABC):
+    """
+    Interface for a Stage in the ExecuTorch export pipeline.
+
+    Each stage can be connected to other stages to form a pipeline.
+    Stages have clear run and get_outputs functions to make the data flow explicit.
+    Each stage implements its own run method with specific parameter names.
+    """
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """
+        Returns the name of this stage.
+        """
+        pass
+
+    @abstractmethod
+    def run(self, **kwargs) -> None:
+        """
+        Executes this stage with the given inputs.
+
+        Each concrete stage class implements this method with specific parameter names.
+        """
+        pass
+
+    @abstractmethod
+    def get_outputs(self) -> Any:
+        """
+        Returns the outputs generated by this stage.
+
+        Returns:
+            The outputs of this stage, to be used as inputs for the next stage
+        """
+        pass
+
+
+class ExportStage(Stage):
+    """
+    First stage: Export PyTorch model to ExportedProgram.
+    """
+
+    def __init__(
+        self,
+        pre_edge_transform_passes: Optional[
+            Callable[[ExportedProgram], ExportedProgram]
+        ] = None,
+    ) -> None:
+        self._exported_program: Dict[str, ExportedProgram] = {}
+        self._pre_edge_transform_passes = pre_edge_transform_passes
+        self._model_dict: Dict[str, nn.Module] = {}
+        self._example_inputs_dict: Dict[str, List[tuple[torch.Tensor, ...]]] = {}
+        self._dynamic_shapes_dict: Dict[str, Any] = {}
+
+    @property
+    def name(self) -> str:
+        return "export"
+
+    def run(
+        self,
+        models: Dict[str, Any],
+        export_config: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ) -> None:
+        """
+        Export PyTorch model to ExportedProgram.
+
+        Args:
+            models: Dictionary mapping method names to PyTorch models
+            export_config: Configuration containing example inputs and dynamic shapes
+            **kwargs: Additional keyword arguments (not used)
+        """
+        # Store inputs
+        self._model_dict = models.get("model", {})
+
+        if export_config is not None:
+            self._example_inputs_dict = export_config.get("example_inputs", {})
+            self._dynamic_shapes_dict = export_config.get("dynamic_shapes", {})
+
+        # Check if we need to do export (if _exported_program is empty)
+        if not self._exported_program:
```

Comment:
What if you want to rexport? For different input shapes or export args? If ExportStange is strictly for one export session and to export with different args you need a new ExportStage than I will make it immutable.

Reply:
This wasn't needed, removed it. One export session is intended to be only for one set of input args.